questdb-rs/src/ingress/ndarr.rs

```diff
@@ -0,0 +1,133 @@
+pub(crate) const MAX_DIMS: usize = 32;
+
+pub trait NdArrayView<T>
+where
+    T: ArrayElement,
+{
+    /// Returns the number of dimensions (rank) of the array.
+    fn ndim(&self) -> usize;
+
+    /// Returns the size of the specified dimension.
+    fn dim(&self, index: usize) -> Option<usize>;
+
+    /// Writes array data to buffer in row-major order.
+    ///
+    /// # Important Notes
+    /// - Buffer must be pre-allocated with exact required size
+    /// - No alignment assumptions should be made about buffer start
+    /// - Handles both contiguous and non-contiguous memory layouts
+    fn write_row_major_buf(&self, buff: &mut [u8]);
+}
+
+/// Marker trait for valid array element types.
+///
+/// Implemented for primitive types that can be stored in arrays.
+/// Combines type information with data type classification.
+pub trait ArrayElement: Copy + 'static {
+    /// Returns the corresponding data type classification.
+    ///
+    /// This enables runtime type identification while maintaining
+    /// compile-time type safety.
+    fn elem_type() -> ElemDataType;
+}
+
+#[repr(u8)]
+#[derive(Debug, PartialEq)]
+pub enum ElemDataType {
```

Comment:
Where does this come from?
Is there a spec document that maps the supported datatypes?

If there's a format document link to it, otherwise let's write a spec for it and link to it.
This enum here is just too confusing and too error-prone otherwise.

Each entry should also be explicit about its value, e.g. `Boolean = 1`, `Byte = 2` .. etc.

Additionally, why is there an `Undefined` entry?

Reply:
PS, do we really intend to support all of these datatypes as binary?

Reply:
Then this list should only contain double :-)